File: bldr/web/boot/validator.test.ts
import { readFileSync } from 'node:fs'
import { resolve } from 'node:path'

import { describe, expect, it, vi } from 'vitest'

import {
  BootReport,
  BootValidation,
  BootValidationViolationKind,
} from './report.pb.js'
import { marshalBootReportJson } from './codec.js'
import { validateBootReport } from './validator.js'

function loadBootReportFixture(name: string) {
  return BootReport.fromJsonString(
    readFileSync(
      resolve(process.cwd(), `bldr/web/boot/testdata/${name}.json`),
      'utf8',
    ),
  )
}

function cloneBootReportFixture(name: string) {
  const report = BootReport.clone(loadBootReportFixture(name))
  if (!report) throw new Error(`failed to clone ${name} fixture`)
  return report
}

describe('validateBootReport', () => {
  const tests = [
    { name: 'successful', pass: true, kinds: [] },
    { name: 'failed', pass: true, kinds: [] },
    { name: 'aborted', pass: true, kinds: [] },
    { name: 'concurrent', pass: true, kinds: [] },
    {
      name: 'reordered',
      pass: false,
      kinds: [BootValidationViolationKind.MARK_ORDER],
    },
    {
      name: 'stalled',
      pass: false,
      kinds: [
        BootValidationViolationKind.GAP_COVERAGE,
        BootValidationViolationKind.GENERIC_LEAF,
        BootValidationViolationKind.UNKNOWN_SHARE,
      ],
    },
    {
      name: 'actionable-ancestor',
      pass: false,
      kinds: [
        BootValidationViolationKind.GENERIC_LEAF,
        BootValidationViolationKind.GENERIC_LEAF,
        BootValidationViolationKind.UNKNOWN_SHARE,
      ],
    },
    {
      name: 'duplicate-span',
      pass: false,
      kinds: [BootValidationViolationKind.SPAN_CONTRACT],
    },
    {
      name: 'deep-span',
      pass: false,
      kinds: [BootValidationViolationKind.SPAN_CONTRACT],
    },
    {
      name: 'post-terminal',
      pass: false,
      kinds: [BootValidationViolationKind.TERMINAL_CONTRACT],
    },
    {
      name: 'invalid-privacy',
      pass: false,
      kinds: [BootValidationViolationKind.TERMINAL_CONTRACT],
    },
    {
      name: 'invalid-accounting',
      pass: false,
      kinds: [BootValidationViolationKind.REPORT_CONTRACT],
    },
    {
      name: 'nonfinite-scalar',
      pass: false,
      kinds: [BootValidationViolationKind.REPORT_CONTRACT],
    },
    {
      name: 'oversized-record',
      pass: false,
      kinds: [BootValidationViolationKind.REPORT_CONTRACT],
    },
  ] as const

  for (const test of tests) {
    it(`validates the ${test.name} golden report`, () => {
      const report = loadBootReportFixture(test.name)
      const validation = validateBootReport(report)
      expect(BootValidation.toJsonString(validation)).toBe(
        BootValidation.toJsonString(report.validation ?? {}),
      )
      expect(validation.pass).toBe(test.pass)
      expect(validation.violations?.map((entry) => entry.kind)).toEqual(
        test.kinds,
      )
    })
  }

  it('rejects multiple usable marks', () => {
    const report = cloneBootReportFixture('successful')
    const marks = report.marks ?? []
    marks[1].label = report.usableMark
    const validation = validateBootReport(report)
    expect(validation.violations?.map((entry) => entry.kind)).toContain(
      BootValidationViolationKind.TERMINAL_CONTRACT,
    )
  })

  it('rejects the shared forbidden privacy vocabulary', () => {
    const values = readFileSync(
      resolve(process.cwd(), 'bldr/web/boot/testdata/privacy-vectors.txt'),
      'utf8',
    )
      .trim()
      .split('\n')
    for (const value of values) {
      const report = cloneBootReportFixture('failed')
      report.terminalErrorCode = value
      const validation = validateBootReport(report)
      expect(validation.violations?.map((entry) => entry.kind)).toContain(
        BootValidationViolationKind.TERMINAL_CONTRACT,
      )
    }
  })

  it('rejects collection and ordered-detail bounds', () => {
    const oversized = cloneBootReportFixture('successful')
    oversized.marks = Array.from({ length: 4097 }, () => ({}))
    expect(
      validateBootReport(oversized).violations?.map((entry) => entry.kind),
    ).toContain(BootValidationViolationKind.REPORT_CONTRACT)

    const unordered = cloneBootReportFixture('successful')
    const detail = unordered.marks?.[0]?.detail
    if (detail) detail.reverse()
    expect(
      validateBootReport(unordered).violations?.map((entry) => entry.kind),
    ).toContain(BootValidationViolationKind.REPORT_CONTRACT)
  })

  it('round trips generated binary and proto JSON', () => {
    const report = loadBootReportFixture('successful')
    const binary = BootReport.toBinary(report)
    const binaryDecoded = BootReport.fromBinary(binary)
    expect(BootReport.toBinary(binaryDecoded)).toEqual(binary)
    const json = BootReport.toJsonString(binaryDecoded)
    const jsonDecoded = BootReport.fromJsonString(json)
    expect(BootReport.toJsonString(jsonDecoded)).toBe(json)
  })

  it('matches the repeated Go and TypeScript canonical proto JSON golden', () => {
    const report = loadBootReportFixture('successful')
    const golden = readFileSync(
      resolve(
        process.cwd(),
        'bldr/web/boot/testdata/successful.canonical.json',
      ),
      'utf8',
    )
    expect(BootReport.toBinary(BootReport.fromJsonString(golden))).toEqual(
      BootReport.toBinary(report),
    )
    const compact = JSON.stringify(JSON.parse(golden))
    for (let attempt = 0; attempt < 20; attempt++) {
      expect(marshalBootReportJson(report)).toBe(compact)
    }
  })

  it('applies the shared field-specific vocabulary vectors', () => {
    const vectors = readFileSync(
      resolve(process.cwd(), 'bldr/web/boot/testdata/vocabulary-vectors.txt'),
      'utf8',
    )
      .trim()
      .split('\n')
    const kinds = {
      REPORT_CONTRACT: BootValidationViolationKind.REPORT_CONTRACT,
      SPAN_CONTRACT: BootValidationViolationKind.SPAN_CONTRACT,
    } as const
    for (const vector of vectors) {
      const [disposition, target, value, expected] = vector.split('|')
      const report = cloneBootReportFixture('successful')
      if (target === 'report-id') report.reportId = value
      else if (target === 'mark-label') report.marks![1].label = value
      else if (target === 'detail-string') {
        report.marks![0].detail![1].value = {
          value: { case: 'stringValue', value },
        }
      } else if (target === 'operation') report.spans![1].operation = value
      else if (target === 'browser-engine') report.environment!.browserEngine = value
      else if (target === 'os-family') report.environment!.osFamily = value
      else if (target === 'worker-mode') report.build!.workerMode = Number(value)
      else throw new Error(`unknown target ${target}`)
      const validation = validateBootReport(report)
      if (disposition === 'accept') expect(validation.pass).toBe(true)
      else {
        expect(validation.violations?.map((entry) => entry.kind)).toContain(
          kinds[expected as keyof typeof kinds],
        )
      }
    }
  })

  it('rejects every shared unsupported enum vector after binary decoding', () => {
    const vectors = readFileSync(
      resolve(process.cwd(), 'bldr/web/boot/testdata/unsupported-enums.txt'),
      'utf8',
    )
      .trim()
      .split('\n')
    const kinds = {
      REPORT_CONTRACT: BootValidationViolationKind.REPORT_CONTRACT,
      TERMINAL_CONTRACT: BootValidationViolationKind.TERMINAL_CONTRACT,
      MARK_ORDER: BootValidationViolationKind.MARK_ORDER,
      SPAN_CONTRACT: BootValidationViolationKind.SPAN_CONTRACT,
    } as const
    for (const vector of vectors) {
      const [target, rawValue, expected] = vector.split('|')
      const value = Number(rawValue)
      const report = cloneBootReportFixture('successful')
      switch (target) {
        case 'report-state':
          report.state = value
          break
        case 'build-type':
          report.build!.buildType = value
          break
        case 'runtime-kind':
          report.build!.runtimeKind = value
          break
        case 'worker-mode':
          report.build!.workerMode = value
          break
        case 'environment-class':
          report.environment!.class = value
          break
        case 'service-worker-state':
          report.environment!.serviceWorkerState = value
          break
        case 'cache-state':
          report.environment!.cacheState = value
          break
        case 'recovery-decision':
          report.environment!.recoveryDecision = value
          break
        case 'mark-phase':
          report.marks![0].phase = value
          break
        case 'counter-unit':
          report.accounting!.samples![0].unit = value
          break
        case 'attachment-kind':
          report.attachments = [
            {
              artifactId: 'boot-artifact-fixture',
              kind: value,
              contentHash: 'a'.repeat(64),
              sizeBytes: 1n,
              releaseGeneration: 'fixture',
            },
          ]
          break
        case 'share-destination':
          report.privacy!.sharedUnixMicros = 1n
          report.privacy!.shareDestination = value
          break
        case 'span-result':
          report.spans![0].result = value
          break
        case 'span-work-class':
          report.spans![0].workClass = value
          break
        default:
          throw new Error(`unknown enum target ${target}`)
      }
      const decoded = BootReport.fromBinary(BootReport.toBinary(report))
      expect(
        validateBootReport(decoded).violations?.map((entry) => entry.kind),
      ).toContain(kinds[expected as keyof typeof kinds])
    }
  })

  it('returns before traversing an oversized graph', () => {
    const report = cloneBootReportFixture('successful')
    report.spans = new Proxy(
      Array.from({ length: 4097 }, () => ({})),
      {
        get(target, property, receiver) {
          if (property === Symbol.iterator) {
            throw new Error('oversized spans were traversed')
          }
          return Reflect.get(target, property, receiver)
        },
      },
    )
    expect(
      validateBootReport(report).violations?.map((entry) => entry.kind),
    ).toEqual([BootValidationViolationKind.REPORT_CONTRACT])
  })

  it('validates without materializing the complete protobuf', () => {
    const toBinary = vi.spyOn(BootReport, 'toBinary').mockImplementation(() => {
      throw new Error('complete protobuf was materialized')
    })
    try {
      expect(validateBootReport(loadBootReportFixture('successful')).pass).toBe(
        true,
      )
      const oversized = cloneBootReportFixture('successful')
      oversized.entrypointId = 'a'.repeat(129)
      expect(
        validateBootReport(oversized).violations?.map((entry) => entry.kind),
      ).toContain(BootValidationViolationKind.REPORT_CONTRACT)
    } finally {
      toBinary.mockRestore()
    }
  })
})

describe('review regressions', () => {
  function expectReportContract(report: BootReport) {
    expect(
      validateBootReport(report).violations?.map((entry) => entry.kind),
    ).toContain(BootValidationViolationKind.REPORT_CONTRACT)
  }

  it('rejects oversized reports before invalid vocabulary returns', () => {
    const report = cloneBootReportFixture('successful')
    report.terminalErrorCode = 'x'.repeat(4 << 20)
    expectReportContract(report)
  })

  it('rejects out-of-range signed and unsigned scalars', () => {
    for (const mutate of [
      (report: BootReport) => {
        report.marks![0].detail![0].value = {
          value: { case: 'unsignedValue', value: -1n },
        }
      },
      (report: BootReport) => {
        report.marks![0].detail![0].value = {
          value: { case: 'signedValue', value: 1n << 63n },
        }
      },
      (report: BootReport) => {
        report.accounting!.samples![0].monotonicMicros = -1n
      },
    ]) {
      const report = cloneBootReportFixture('successful')
      mutate(report)
      expectReportContract(report)
    }
  })

  it('enforces the shared timestamp int64 boundary', () => {
    const oversized = cloneBootReportFixture('successful')
    oversized.privacy!.sharedUnixMicros = 1n << 63n
    oversized.privacy!.shareDestination = 1
    expectReportContract(oversized)

    const maximum = cloneBootReportFixture('successful')
    maximum.privacy!.sharedUnixMicros = (1n << 63n) - 1n
    maximum.privacy!.shareDestination = 1
    expect(validateBootReport(maximum).pass).toBe(true)
  })

  it('accepts numeric identifier suffixes', () => {
    const report = cloneBootReportFixture('successful')
    report.reportId = 'boot-report-1'
    report.spans![0].spanId = 'span-1'
    for (const span of report.spans!.slice(1)) {
      if (span.parentSpanId === 'runtime') span.parentSpanId = 'span-1'
    }
    report.attachments = [
      {
        artifactId: 'boot-artifact-1',
        kind: 1,
        contentHash: 'a'.repeat(64),
        sizeBytes: 1n,
        releaseGeneration: report.build!.releaseGeneration,
      },
    ]
    expect(validateBootReport(report).pass).toBe(true)
  })

  it('ignores stale stored validation enum metadata', () => {
    const report = cloneBootReportFixture('successful')
    report.validation!.violations = [
      { kind: 99 as BootValidationViolationKind },
    ]
    expect(validateBootReport(report).pass).toBe(true)
  })

  it('rejects failed and aborted reports containing the usable mark', () => {
    for (const name of ['failed', 'aborted']) {
      const report = cloneBootReportFixture(name)
      report.marks![0].label = report.usableMark
      expect(
        validateBootReport(report).violations?.map((entry) => entry.kind),
      ).toContain(BootValidationViolationKind.TERMINAL_CONTRACT)
    }
  })

  it('requires accounting samples at exact mark boundaries', () => {
    const report = cloneBootReportFixture('successful')
    report.accounting!.samples![0].monotonicMicros = 1n
    expectReportContract(report)
  })

  it('requires attachment and report release generations to match', () => {
    const report = cloneBootReportFixture('successful')
    report.attachments = [
      {
        artifactId: 'boot-artifact-1',
        kind: 1,
        contentHash: 'a'.repeat(64),
        sizeBytes: 1n,
        releaseGeneration: 'other',
      },
    ]
    expectReportContract(report)
  })
})

File: bldr/web/boot/validator_test.go
package boot

import (
	"bytes"
	"encoding/json" //nolint:depguard // json.Compact canonicalizes formatter whitespace in a test fixture.
	"os"
	"path/filepath"
	"reflect"
	"slices"
	"strconv"
	"strings"
	"testing"
)

func loadBootReportFixture(t *testing.T, name string) *BootReport {
	t.Helper()
	data, err := os.ReadFile(filepath.Join("testdata", name+".json"))
	if err != nil {
		t.Fatal(err)
	}
	report := new(BootReport)
	if err := report.UnmarshalJSON(data); err != nil {
		t.Fatal(err)
	}
	return report
}

func bootViolationKinds(validation *BootValidation) []BootValidationViolationKind {
	kinds := make([]BootValidationViolationKind, 0, len(validation.GetViolations()))
	for _, violation := range validation.GetViolations() {
		kinds = append(kinds, violation.GetKind())
	}
	return kinds
}

func TestValidateBootReportGoldenReports(t *testing.T) {
	tests := []struct {
		name  string
		pass  bool
		kinds []BootValidationViolationKind
	}{
		{name: "successful", pass: true},
		{name: "failed", pass: true},
		{name: "aborted", pass: true},
		{name: "concurrent", pass: true},
		{name: "reordered", kinds: []BootValidationViolationKind{
			BootValidationViolationKind_BOOT_VALIDATION_VIOLATION_KIND_MARK_ORDER,
		}},
		{name: "stalled", kinds: []BootValidationViolationKind{
			BootValidationViolationKind_BOOT_VALIDATION_VIOLATION_KIND_GAP_COVERAGE,
			BootValidationViolationKind_BOOT_VALIDATION_VIOLATION_KIND_GENERIC_LEAF,
			BootValidationViolationKind_BOOT_VALIDATION_VIOLATION_KIND_UNKNOWN_SHARE,
		}},
		{name: "actionable-ancestor", kinds: []BootValidationViolationKind{
			BootValidationViolationKind_BOOT_VALIDATION_VIOLATION_KIND_GENERIC_LEAF,
			BootValidationViolationKind_BOOT_VALIDATION_VIOLATION_KIND_GENERIC_LEAF,
			BootValidationViolationKind_BOOT_VALIDATION_VIOLATION_KIND_UNKNOWN_SHARE,
		}},
		{name: "duplicate-span", kinds: []BootValidationViolationKind{
			BootValidationViolationKind_BOOT_VALIDATION_VIOLATION_KIND_SPAN_CONTRACT,
		}},
		{name: "deep-span", kinds: []BootValidationViolationKind{
			BootValidationViolationKind_BOOT_VALIDATION_VIOLATION_KIND_SPAN_CONTRACT,
		}},
		{name: "post-terminal", kinds: []BootValidationViolationKind{
			BootValidationViolationKind_BOOT_VALIDATION_VIOLATION_KIND_TERMINAL_CONTRACT,
		}},
		{name: "invalid-privacy", kinds: []BootValidationViolationKind{
			BootValidationViolationKind_BOOT_VALIDATION_VIOLATION_KIND_TERMINAL_CONTRACT,
		}},
		{name: "invalid-accounting", kinds: []BootValidationViolationKind{
			BootValidationViolationKind_BOOT_VALIDATION_VIOLATION_KIND_REPORT_CONTRACT,
		}},
		{name: "nonfinite-scalar", kinds: []BootValidationViolationKind{
			BootValidationViolationKind_BOOT_VALIDATION_VIOLATION_KIND_REPORT_CONTRACT,
		}},
		{name: "oversized-record", kinds: []BootValidationViolationKind{
			BootValidationViolationKind_BOOT_VALIDATION_VIOLATION_KIND_REPORT_CONTRACT,
		}},
	}
	for _, test := range tests {
		t.Run(test.name, func(t *testing.T) {
			report := loadBootReportFixture(t, test.name)
			validation := ValidateBootReport(report)
			if !reflect.DeepEqual(validation, report.GetValidation()) {
				t.Fatalf("validation differs from shared golden: got %v, want %v", validation, report.GetValidation())
			}
			if validation.GetPass() != test.pass {
				t.Fatalf("pass = %v, violations = %v", validation.GetPass(), validation.GetViolations())
			}
			if kinds := bootViolationKinds(validation); !slices.Equal(kinds, test.kinds) {
				t.Fatalf("violation kinds = %v, want %v", kinds, test.kinds)
			}
		})
	}
}

func TestBootReportGeneratedCodeRoundTrip(t *testing.T) {
	report := loadBootReportFixture(t, "successful")
	binary, err := report.MarshalVT()
	if err != nil {
		t.Fatal(err)
	}
	decoded := new(BootReport)
	if err := decoded.UnmarshalVT(binary); err != nil {
		t.Fatal(err)
	}
	if !reflect.DeepEqual(decoded, report) {
		t.Fatal("binary round trip changed report")
	}
	jsonData, err := decoded.MarshalJSON()
	if err != nil {
		t.Fatal(err)
	}
	jsonDecoded := new(BootReport)
	if err := jsonDecoded.UnmarshalJSON(jsonData); err != nil {
		t.Fatal(err)
	}
	if !reflect.DeepEqual(jsonDecoded, report) {
		t.Fatal("proto JSON round trip changed report")
	}
}

func TestMarshalBootReportJSONMatchesCrossLanguageGolden(t *testing.T) {
	report := loadBootReportFixture(t, "successful")
	golden, err := os.ReadFile(filepath.Join("testdata", "successful.canonical.json"))
	if err != nil {
		t.Fatal(err)
	}
	goldenReport := new(BootReport)
	if err := goldenReport.UnmarshalJSON(golden); err != nil {
		t.Fatal(err)
	}
	if !reflect.DeepEqual(goldenReport, report) {
		t.Fatal("canonical proto JSON changed report semantics")
	}
	var compact bytes.Buffer
	if err := json.Compact(&compact, golden); err != nil {
		t.Fatal(err)
	}
	for range 20 {
		encoded, err := MarshalBootReportJSON(report)
		if err != nil {
			t.Fatal(err)
		}
		if !slices.Equal(encoded, compact.Bytes()) {
			t.Fatalf("canonical proto JSON differs:\n got %s\nwant %s", encoded, compact.Bytes())
		}
	}
}

func TestValidateBootReportRejectsMultipleUsableMarks(t *testing.T) {
	report := loadBootReportFixture(t, "successful").CloneVT()
	report.Marks[1].Label = report.GetUsableMark()
	validation := ValidateBootReport(report)
	if !slices.Contains(bootViolationKinds(validation),
		BootValidationViolationKind_BOOT_VALIDATION_VIOLATION_KIND_TERMINAL_CONTRACT) {
		t.Fatalf("violations = %v, want terminal contract", validation.GetViolations())
	}
}

func TestValidateBootReportRejectsPrivacyVocabulary(t *testing.T) {
	data, err := os.ReadFile(filepath.Join("testdata", "privacy-vectors.txt"))
	if err != nil {
		t.Fatal(err)
	}
	for value := range strings.SplitSeq(strings.TrimSpace(string(data)), "\n") {
		t.Run(value, func(t *testing.T) {
			report := loadBootReportFixture(t, "failed").CloneVT()
			report.TerminalErrorCode = value
			validation := ValidateBootReport(report)
			if !slices.Contains(bootViolationKinds(validation),
				BootValidationViolationKind_BOOT_VALIDATION_VIOLATION_KIND_TERMINAL_CONTRACT) {
				t.Fatalf("violations = %v, want terminal contract", validation.GetViolations())
			}
		})
	}
}

func TestValidateBootReportRejectsCollectionAndDetailBounds(t *testing.T) {
	report := loadBootReportFixture(t, "successful").CloneVT()
	report.Marks = make([]*BootMark, maxBootMarks+1)
	if !slices.Contains(bootViolationKinds(ValidateBootReport(report)),
		BootValidationViolationKind_BOOT_VALIDATION_VIOLATION_KIND_REPORT_CONTRACT) {
		t.Fatal("oversized marks did not fail the report contract")
	}

	report = loadBootReportFixture(t, "successful").CloneVT()
	report.Marks[0].Detail[0], report.Marks[0].Detail[1] = report.Marks[0].Detail[1], report.Marks[0].Detail[0]
	if !slices.Contains(bootViolationKinds(ValidateBootReport(report)),
		BootValidationViolationKind_BOOT_VALIDATION_VIOLATION_KIND_REPORT_CONTRACT) {
		t.Fatal("unordered detail did not fail the report contract")
	}
}

func TestValidateBootReportSharedVocabularyVectors(t *testing.T) {
	data, err := os.ReadFile(filepath.Join("testdata", "vocabulary-vectors.txt"))
	if err != nil {
		t.Fatal(err)
	}
	for line := range strings.SplitSeq(strings.TrimSpace(string(data)), "\n") {
		parts := strings.Split(line, "|")
		t.Run(parts[0]+"-"+parts[1]+"-"+parts[2], func(t *testing.T) {
			report := loadBootReportFixture(t, "successful").CloneVT()
			switch parts[1] {
			case "report-id":
				report.ReportId = parts[2]
			case "mark-label":
				report.Marks[1].Label = parts[2]
			case "detail-string":
				report.Marks[0].Detail[1].Value.Value = &BootValue_StringValue{StringValue: parts[2]}
			case "operation":
				report.Spans[1].Operation = parts[2]
			case "browser-engine":
				report.Environment.BrowserEngine = parts[2]
			case "os-family":
				report.Environment.OsFamily = parts[2]
			case "worker-mode":
				mode, err := strconv.Atoi(parts[2])
				if err != nil {
					t.Fatal(err)
				}
				report.Build.WorkerMode = BootWorkerMode(mode)
			default:
				t.Fatalf("unknown target %q", parts[1])
			}
			validation := ValidateBootReport(report)
			if parts[0] == "accept" {
				if !validation.GetPass() {
					t.Fatalf("violations = %v, want pass", validation.GetViolations())
				}
				return
			}
			want := map[string]BootValidationViolationKind{
				"REPORT_CONTRACT": BootValidationViolationKind_BOOT_VALIDATION_VIOLATION_KIND_REPORT_CONTRACT,
				"SPAN_CONTRACT":   BootValidationViolationKind_BOOT_VALIDATION_VIOLATION_KIND_SPAN_CONTRACT,
			}[parts[3]]
			if !slices.Contains(bootViolationKinds(validation), want) {
				t.Fatalf("violations = %v, want %v", validation.GetViolations(), want)
			}
		})
	}
}

func TestValidateBootReportRejectsSharedUnsupportedEnumVectorsFromBinary(t *testing.T) {
	data, err := os.ReadFile(filepath.Join("testdata", "unsupported-enums.txt"))
	if err != nil {
		t.Fatal(err)
	}
	for line := range strings.SplitSeq(strings.TrimSpace(string(data)), "\n") {
		parts := strings.Split(line, "|")
		t.Run(parts[0], func(t *testing.T) {
			report := loadBootReportFixture(t, "successful").CloneVT()
			switch parts[0] {
			case "report-state":
				report.State = BootReportState(99)
			case "build-type":
				report.Build.BuildType = BootBuildType(99)
			case "runtime-kind":
				report.Build.RuntimeKind = BootRuntimeKind(99)
			case "worker-mode":
				report.Build.WorkerMode = BootWorkerMode(99)
			case "environment-class":
				report.Environment.Class = BootEnvironmentClass(99)
			case "service-worker-state":
				report.Environment.ServiceWorkerState = BootServiceWorkerState(99)
			case "cache-state":
				report.Environment.CacheState = BootCacheState(99)
			case "recovery-decision":
				report.Environment.RecoveryDecision = BootRecoveryDecision(99)
			case "mark-phase":
				report.Marks[0].Phase = BootPhase(99)
			case "counter-unit":
				report.Accounting.Samples[0].Unit = BootCounterUnit(99)
			case "attachment-kind":
				report.Attachments = []*BootAttachment{{
					ArtifactId: "fixture-artifact", Kind: BootAttachmentKind(99),
					ContentHash: strings.Repeat("a", 64), SizeBytes: 1, ReleaseGeneration: "fixture",
				}}
			case "share-destination":
				report.Privacy.SharedUnixMicros = 1
				report.Privacy.ShareDestination = BootShareDestination(99)
			case "span-result":
				report.Spans[0].Result = BootSpanResult(99)
			case "span-work-class":
				report.Spans[0].WorkClass = BootWorkClass(99)
			default:
				t.Fatalf("unknown enum target %q", parts[0])
			}
			binary, marshalErr := report.MarshalVT()
			if marshalErr != nil {
				t.Fatal(marshalErr)
			}
			decoded := new(BootReport)
			if unmarshalErr := decoded.UnmarshalVT(binary); unmarshalErr != nil {
				t.Fatal(unmarshalErr)
			}
			want := map[string]BootValidationViolationKind{
				"REPORT_CONTRACT":   BootValidationViolationKind_BOOT_VALIDATION_VIOLATION_KIND_REPORT_CONTRACT,
				"TERMINAL_CONTRACT": BootValidationViolationKind_BOOT_VALIDATION_VIOLATION_KIND_TERMINAL_CONTRACT,
				"MARK_ORDER":        BootValidationViolationKind_BOOT_VALIDATION_VIOLATION_KIND_MARK_ORDER,
				"SPAN_CONTRACT":     BootValidationViolationKind_BOOT_VALIDATION_VIOLATION_KIND_SPAN_CONTRACT,
			}[parts[2]]
			if kinds := bootViolationKinds(ValidateBootReport(decoded)); !slices.Contains(kinds, want) {
				t.Fatalf("violations = %v, want %v", kinds, want)
			}
		})
	}
}

func TestValidateBootReportReturnsBeforeOversizedGraphWork(t *testing.T) {
	report := loadBootReportFixture(t, "successful").CloneVT()
	report.Spans = make([]*BootSpan, maxBootSpans+1)
	want := []BootValidationViolationKind{
		BootValidationViolationKind_BOOT_VALIDATION_VIOLATION_KIND_REPORT_CONTRACT,
	}
	if kinds := bootViolationKinds(ValidateBootReport(report)); !slices.Equal(kinds, want) {
		t.Fatalf("violations = %v, want early %v", kinds, want)
	}
}

func TestValidateBootReportReviewRegressions(t *testing.T) {
	hasReportContract := func(report *BootReport) bool {
		return slices.Contains(bootViolationKinds(ValidateBootReport(report)),
			BootValidationViolationKind_BOOT_VALIDATION_VIOLATION_KIND_REPORT_CONTRACT)
	}

	t.Run("oversized invalid vocabulary", func(t *testing.T) {
		report := loadBootReportFixture(t, "successful").CloneVT()
		report.TerminalErrorCode = strings.Repeat("x", maxBootRecordBytes)
		if !hasReportContract(report) {
			t.Fatal("oversized report bypassed the report contract")
		}
	})

	t.Run("numeric identifier suffixes", func(t *testing.T) {
		report := loadBootReportFixture(t, "successful").CloneVT()
		report.ReportId = "boot-report-1"
		report.Spans[0].SpanId = "span-1"
		for _, span := range report.Spans[1:] {
			if span.ParentSpanId == "runtime" {
				span.ParentSpanId = "span-1"
			}
		}
		report.Attachments = []*BootAttachment{{
			ArtifactId: "boot-artifact-1", Kind: BootAttachmentKind_BOOT_ATTACHMENT_KIND_SOURCE_MAP,
			ContentHash: strings.Repeat("a", 64), SizeBytes: 1, ReleaseGeneration: report.Build.ReleaseGeneration,
		}}
		if validation := ValidateBootReport(report); !validation.GetPass() {
			t.Fatalf("numeric suffixes rejected: %v", validation.GetViolations())
		}
	})

	t.Run("stored validation metadata ignored", func(t *testing.T) {
		report := loadBootReportFixture(t, "successful").CloneVT()
		report.Validation.Violations = []*BootValidationViolation{{Kind: BootValidationViolationKind(99)}}
		if validation := ValidateBootReport(report); !validation.GetPass() {
			t.Fatalf("stored validation changed derived result: %v", validation.GetViolations())
		}
	})

	t.Run("failed usable mark", func(t *testing.T) {
		report := loadBootReportFixture(t, "failed").CloneVT()
		report.Marks[0].Label = report.UsableMark
		if !slices.Contains(bootViolationKinds(ValidateBootReport(report)),
			BootValidationViolationKind_BOOT_VALIDATION_VIOLATION_KIND_TERMINAL_CONTRACT) {
			t.Fatal("failed report with usable mark passed terminal contract")
		}
	})

	t.Run("accounting time between marks", func(t *testing.T) {
		report := loadBootReportFixture(t, "successful").CloneVT()
		report.Accounting.Samples[0].MonotonicMicros = 1
		if !hasReportContract(report) {
			t.Fatal("accounting sample between mark boundaries passed")
		}
	})

	t.Run("attachment generation mismatch", func(t *testing.T) {
		report := loadBootReportFixture(t, "successful").CloneVT()
		report.Attachments = []*BootAttachment{{
			ArtifactId: "boot-artifact-1", Kind: BootAttachmentKind_BOOT_ATTACHMENT_KIND_SOURCE_MAP,
			ContentHash: strings.Repeat("a", 64), SizeBytes: 1, ReleaseGeneration: "other",
		}}
		if !hasReportContract(report) {
			t.Fatal("attachment from another release generation passed")
		}
	})
}
